server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java

```diff
@@ -345,9 +394,41 @@ public static boolean tryFastFilterAggregation(
         final BiConsumer<Long, Integer> incrementDocCount
     ) throws IOException {
         if (fastFilterContext == null) return false;
-        if (fastFilterContext.filters == null) return false;
+        if (!fastFilterContext.rewriteable || !fastFilterContext.shouldBuildFiltersAtSegmentLevel) {
+            return false;
+        }
+
+        NumericDocValues docCountValues = DocValues.getNumeric(ctx.reader(), DocCountFieldMapper.NAME);
+        if (docCountValues.nextDoc() != NO_MORE_DOCS) {
+            logger.debug(
+                "Shard {} segment {} has at least one document with _doc_count field, skip fast filter optimization",
+                fastFilterContext.context.indexShard().shardId(),
+                ctx.ord
+            );
+            return false;
+        }
+
+        // if no filters built at shard level (see getDateHistoAggBounds method for possible reasons)
+        // check if the query is functionally match-all at segment level
+        if (!fastFilterContext.filtersBuiltAtShardLevel && !segmentMatchAll(fastFilterContext.context, ctx)) {
+            return false;
+        }
+        Weight[] filters = fastFilterContext.filters;
+        if (filters == null) {
+            logger.debug(
+                "Shard {} segment {} functionally match all documents. Build the fast filter",
+                fastFilterContext.context.indexShard().shardId(),
+                ctx.ord
+            );
+            filters = fastFilterContext.buildFastFilter(FastFilterRewriteHelper::getDateHistoAggBoundsSegLevel);
+            if (filters == null) {
+                // At segment level, build filter should only be called once
+                // since the conditions for build filter won't change for other segments
+                fastFilterContext.shouldBuildFiltersAtSegmentLevel = false;
```

Comment:
Doesn't this mean that the logic at line 397:
```
        if (!fastFilterContext.rewriteable || !fastFilterContext.shouldBuildFiltersAtSegmentLevel) {
            return false;
        }
```

will always return `false`? As far as I can tell, the whole method will only execute for a single segment. What if I have some segments that don't benefit from the optimization, but multiple segments that do?

Reply:
Follows the old thinking https://github.com/opensearch-project/OpenSearch/pull/12073#discussion_r1480371737

since the bounds is always same, the filters that gonna be created will always be same, so only compute once.

Now the logic is updated to handle each segment bound separately, so removing this check.